### PR TITLE
Bypass load average arg value through equal sign

### PR DESCRIPTION
**Pull request by webknjaz**
##### SUMMARY
TL;DR Using `loadavg` argument of `portage` module causes task to fail.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
portage

##### ANSIBLE VERSION

```
ansible 2.5.0b2
  config file = None
  configured module search path = ['~/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = ~/.pyenv/versions/3.6.4/lib/python3.6/site-packages/ansible
  executable location = ~/.pyenv/versions/3.6.4/bin/ansible
  python version = 3.6.4 (default, Feb 17 2018, 17:11:17) [GCC 7.3.0]
```

##### ADDITIONAL INFORMATION
```
    "stderr_lines": [
        "!!! '--load-average 4.0' is not a valid package atom.",
        "!!! Please check ebuild(5) for full details."
    ],
```

### CODE REVIEW

**Comment by webknjaz** [comment]
cc: @sayap

**Comment by wltjr** [comment]
@webknjaz That is interesting and either undocumented or something is changing. The man page shows

```
       --load-average [LOAD]
              Specifies  that  no  new  builds  should be started if there are
              other builds running and the load average is at  least  LOAD  (a
              floating-point  number).   With  no argument, removes a previous
              load limit.  This option is recommended for use  in  combination
              with  --jobs  in  order  to  avoid  excess load. See make(1) for
              information about analogous options that  should  be  configured
              via MAKEOPTS in make.conf(5).

       -j [JOBS], --jobs[=JOBS]
              Specifies  the  number  of  packages to build simultaneously. If
              this option is given without an argument, emerge will not  limit
              the  number  of  jobs  that can run simultaneously. Also see the
              related --load-average option.  Similarly to  the  --quiet-build
              option,  the  --jobs  option causes all build output to be redi?
              rected to logs.  Note that interactive packages currently  force
              a  setting of --jobs=1. This issue can be temporarily avoided by
              specifying --accept-properties=-interactive.
```

Clearly something is off per your output. I am pretty sure I tested this, highly doubt I committed without testing. I believe something has changed likely in portage. With the man page not updated or something. Based on portage [code usage](https://github.com/gentoo/portage/search?utf8=%E2%9C%93&q=--load-average&type=) . It seems like the jobs option. 

Did it work with your changes? Do you have a copy of the entire emerge line it was attempting? Maybe something else is off.

**Comment by wltjr** [comment]
Ok having tested this a bit seems like a Ansible portage module bug. Something I need to change for sure. It is quoting the argument and value, which causes the issue. It seems  ```--load-average=x.x``` is valid despite not being documented.

Quotes causes problem
```
# emerge -qkv1 "--load-average 1.0" ansible
!!! '--load-average 1.0' is not a valid package atom.
!!! Please check ebuild(5) for full details.

# emerge -qkv1 "--jobs 2" ansible
!!! '--jobs 2' is not a valid package atom.
!!! Please check ebuild(5) for full details.
```

Both those are valid for portage.
```
# emerge -qkv1O --load-average 1.0 ansible
>>> Emerging binary (1 of 1) app-admin/ansible-2.4.3.0.1::os-xtoo

# emerge -qkv1O --jobs 2 ansible
>>> Emerging binary (1 of 1) app-admin/ansible-2.4.3.0.1::os-xtoo
```

As is
```
# emerge -qkv1O --load-average=1.0 ansible
>>> Emerging binary (1 of 1) app-admin/ansible-2.4.3.0.1::os-xtoo

# emerge -qkv1O --jobs=2 ansible
>>> Emerging binary (1 of 1) app-admin/ansible-2.4.3.0.1::os-xtoo
```

The only problem with ```--load-average=x.x```, is the equal would be all the time. You can use ```--load-average``` with no value. Which the module presently does not allow you to do, and needs to be fixed. ```--jobs``` is different, it always requires an argument. Why it has the equal sign.

I think the whole ```--load-average``` handling needs to be fixed for the following;
 - support ```--load-average``` with no argument
 - either pass equal sign when there is a value or prevent quoting

@webknjaz Would you like to take a shot at addressing those 2 issues and redo PR or submit new? I do not think we can proceed as is. It will fix the issue you ran into, but then you still cannot use ```--load-average``` with no arguments. I would assume we fix this for both cases and have fully working support for that argument.

**Comment by webknjaz** [comment]
Oh sure, i'm personally interested in fixing upstream rather than have a patched fork :)

P.S. It would be super useful if you could add comments in "Files" tab right next to the line you refer to. TIA.

**Comment by webknjaz** [comment]
I'm a bit stuck on designing the right public API for these options. Originally I was thinking that it might be possible to allow specifying `True` for option w/o arg, and specifying int/float requirement otherwise for the same option.
Now it looks like this might require additional options like `reset-loadavg`.

File: lib/ansible/modules/packaging/os/portage.py [hunk 1]
```diff
@@ -326,13 +326,20 @@ def emerge_packages(module, packages):
...
-            args.append(arg + str(p[flag]))
+        flag_val = p[flag]
+        if flag_val is None or flag_val is False:
+            continue
+
+        if flag_val is True:
+            args.append(arg)
+            continue
+
+        args.extend((arg, str(flag_val)))
```
Comment (bcoca) [inline lib/ansible/modules/packaging/os/portage.py]: use to_native (module_utils._text) instead of str, it deals with py2/py3 compatiblity

Comment (webknjaz) [reply lib/ansible/modules/packaging/os/portage.py]: @bcoca done, thanks!

File: lib/ansible/modules/packaging/os/portage.py [hunk 2]
```diff
@@ -326,13 +326,20 @@ def emerge_packages(module, packages):
...
-        'loadavg': '--load-average ',
+        'jobs': '--jobs',
+        'loadavg': '--load-average',
     }
 
     for flag, arg in emerge_flags.items():
-        if p[flag] is not None:
-            args.append(arg + str(p[flag]))
+        flag_val = p[flag]
+        if flag_val is None or flag_val is False:
```
Comment (bcoca) [inline lib/ansible/modules/packaging/os/portage.py]: so the 'is' will only work with boolean values so you cannot set a jobs = 5 nor jobs = '', the original was not completely correct either:

```
if p[flag]:
   args.append(arg + to_native(p[flag])
```
^ this should work for all cases .. except 0 values

Comment (webknjaz) [reply lib/ansible/modules/packaging/os/portage.py]: yes, I just left it copy-pasted for now. will fix

Comment (webknjaz) [reply lib/ansible/modules/packaging/os/portage.py]: I'm still not sure about public API here. Is it possible to make schema accept a truthful boolean or a number in one place?

Comment (bcoca) [reply lib/ansible/modules/packaging/os/portage.py]: yes, type=raw will avoid type enforcement, but it is not normally a good interface

Comment (webknjaz) [reply lib/ansible/modules/packaging/os/portage.py]: What would you advise given that these options can be invoked w/o param?  https://github.com/ansible/ansible/pull/36355#issuecomment-366482579

Comment (bcoca) [reply lib/ansible/modules/packaging/os/portage.py]: probably keep as int and take 0/false values as removing the previous setting.

```python
if val:
   append flag + val
elif val is not None:
   append flag
```
and then just document that in the option, also add 'since version ..' to make sure those with older versions don't think this behaviour applies to them.

Comment (webknjaz) [reply lib/ansible/modules/packaging/os/portage.py]: Will the type checker allow user bypassing `False` into that int value? Or you mean it's okay to introduce special behavior for `int(0)` since it's being `bool`eaned as `False`?

**Comment by webknjaz** [comment]
@bcoca @wltjr I think, I've adjusted it to work as requested.

File: lib/ansible/modules/packaging/os/portage.py [hunk 3]
```diff
@@ -326,13 +331,20 @@ def emerge_packages(module, packages):
         module.fail_json(msg='Use only one of usepkg, usepkgonly')
 
     emerge_flags = {
-        'jobs': '--jobs=',
-        'loadavg': '--load-average ',
+        'jobs': '--jobs',
+        'loadavg': '--load-average',
```
Comment (webknjaz) [inline lib/ansible/modules/packaging/os/portage.py]: turns out, it still doesn't work for `0`, because `0` is turned into `float(0.0)` and then turned into `str(0.0)`.

File: lib/ansible/modules/packaging/os/portage.py [hunk 4]
```diff
@@ -326,13 +331,18 @@ def emerge_packages(module, packages):
...
     }
 
     for flag, arg in emerge_flags.items():
-        if p[flag] is not None:
-            args.append(arg + str(p[flag]))
+        flag_val = p[flag]
+        if not flag_val:
+            if flag_val is not None:
+                args.append(arg)
+            continue
```
Comment (bcoca) [inline lib/ansible/modules/packaging/os/portage.py]: you really dont need the continue

File: lib/ansible/modules/packaging/os/portage.py [hunk 5]
```diff
@@ -326,13 +331,18 @@ def emerge_packages(module, packages):
...
+
+        args.extend((arg, to_native(flag_val)))
```
Comment (bcoca) [inline lib/ansible/modules/packaging/os/portage.py]: not sure if its just me but i still find it confusing, wouldn't it be cleaner as?:

```python
        flag_val = p[flag]
        if flag_val:
           args.extend((arg, to_native(flag_val)))
        elif flag_val is not None:
           args.append(arg)
```

Comment (webknjaz) [reply lib/ansible/modules/packaging/os/portage.py]: I tried to avoid nesting here. Maybe I can reduce it to be one-level.

Comment (webknjaz) [reply lib/ansible/modules/packaging/os/portage.py]: @bcoca It must be cleaner now

File: lib/ansible/modules/packaging/os/portage.py [hunk 6]
```diff
@@ -156,6 +156,8 @@
   jobs:
     description:
       - Specifies the number of packages to build simultaneously.
+      - Since version 2.6: Value of 0 resets any previously added
```
Comment (bcoca) [inline lib/ansible/modules/packaging/os/portage.py]: you need to quote : as it is a 'yaml' delimiter

File: lib/ansible/modules/packaging/os/portage.py (hunk 5 above)
Comment (wltjr) [reply lib/ansible/modules/packaging/os/portage.py]: Thanks for making it clearer and comments. Though I think I prefer bcoca approach. Less code, and does not require comments. I read it as either add the flag and value, or if not none, just add the flag. If neither, it does nothing and onto the next iteration. Does not require any continues. Seems more compact and elegant. But either can work. More what the project prefers, so would defer to bcocas preference.

**Comment by wltjr** [comment]
@bcoca general question, should the commits be merged into a single or left as individual? I think the bot merges into a single. Personally prefer it squashed into one to see all changes at once. Patch 2 some what undoes patch 1 and is not really necessary. Pretty moot, just makes it easier to see the total changes. I am good either way. Just curious as I have not had to much experience handling Ansible PRs with multiple commits. Not seeing info in guides. I myself tend to force push and have it all in a single. Though I normally try to make any commit as small as possible.

@webknjaz nothing for you here, just asking general contributor/maintainer questions. Though if it is preferred to be single vs multiple. Would need to squash. Otherwise I am just asking for my own knowledge.

File: lib/ansible/modules/packaging/os/portage.py (hunk 5 above)
Comment (bcoca) [reply lib/ansible/modules/packaging/os/portage.py]: At the point it is 'clear' i'm fine with either way, i am not going to force a style unless the code need de-obfuscation.

**Comment by wltjr** [comment]
I see [PR.diff](https://patch-diff.githubusercontent.com/raw/ansible/ansible/pull/36355.diff) is better than [PR.patch](https://patch-diff.githubusercontent.com/raw/ansible/ansible/pull/36355.patch), makes it messy.

File: lib/ansible/modules/packaging/os/portage.py (hunk 6 above)
Comment (webknjaz) [reply lib/ansible/modules/packaging/os/portage.py]: Oops. Btw why ain't yaml blocks used here?

File: lib/ansible/modules/packaging/os/portage.py (hunk 5 above)
Comment (wltjr) [reply lib/ansible/modules/packaging/os/portage.py]: Ok I will leave it to the contributor. Rather not debate over stylish stuff that is moot. I am good either way, long as it works. Thanks!

**Comment by bcoca** [comment]
@wltjr we normally prefer 1 commit as it makes cherry picking/revert/etc very simple and as you mention it obscures 'workflow' to final patch. That said, since github now supports a 'squash merge' we don't really make authors sweat about it.

**Comment by wltjr** [comment]
@bcoca now that I am aware I can use diff vs patch. That helps in review. I assumed the squash. Thus asking before requesting anything further of contributors. Its moot now that I can see it all on one via diff. Thanks for confirmation though!

**Comment by wltjr** [comment]
@webknjaz looks good and thank you for your contribution and working the issues raised in this PR. Seems this addresses another issues with ```--jobs```, as it can be used without any arguments. The module did not support such before. In fixing it for ```--load-average```, also indirectly fixed using ```--jobs``` without any options. Also lays the foundation for future flags with no arguments. Good stuff, thanks!

**Comment by wltjr** [comment]
Ok we are close, but no cigar. May need to try @bcoca modifications. Seems something is going wrong in that section. It works for ```--jobs``` when passing zero. But does not work for loadavg. When you set loadavg to zero, you do not get ```--load-average``` at all. Unless I did something wrong, but seems pretty straight forward.

Using the following test playbook
```yaml
---
- name: Test portage changes
  hosts: wlt
  gather_facts: no
  remote_user: root

  tasks:

    - name: Test portage changes
      portage:
        package: [nano, curl]
        getbinpkg: yes
        jobs: 0
        state: latest

    - name: Test portage changes
      portage:
        package: [nano, curl]
        getbinpkg: yes
        jobs: 1
        state: latest

    - name: Test portage changes
      portage:
        package: [nano, curl]
        getbinpkg: yes
        loadavg: 0
        state: latest

    - name: Test portage changes
      portage:
        package: [nano, curl]
        getbinpkg: yes
        loadavg: 2.0
        state: latest

    - name: Test portage changes
      portage:
        package: [nano, curl]
        getbinpkg: yes
        jobs: 0
        loadavg: 0
        state: latest

    - name: Test portage changes
      portage:
        package: [nano, curl]
        getbinpkg: yes
        jobs: 1
        loadavg: 2.0
        state: latest
```

Generated the following output
```
1519064457: Started emerge on: Feb 19, 2018 13:20:57
1519064457:  *** emerge --jobs --update --usepkg --getbinpkg nano curl

1519064521: Started emerge on: Feb 19, 2018 13:22:00
1519064521:  *** emerge --jobs=1 --usepkg --getbinpkg --update nano curl

1519064599: Started emerge on: Feb 19, 2018 13:23:18
1519064599:  *** emerge --update --getbinpkg --usepkg nano curl

1519064675: Started emerge on: Feb 19, 2018 13:24:35
1519064675:  *** emerge --getbinpkg --usepkg --update --load-average=2.0 nano curl

1519064750: Started emerge on: Feb 19, 2018 13:25:50
1519064750:  *** emerge --update --jobs --getbinpkg --usepkg nano curl

1519064839: Started emerge on: Feb 19, 2018 13:27:18
1519064839:  *** emerge --jobs=1 --getbinpkg --update --load-average=2.0 --usepkg nano curl
```

Missing the following
```
emerge --update --getbinpkg --load-average  --usepkg nano curl
emerge  --jobs --update  --getbinpkg --load-average  --usepkg nano curl
```

Not sure why it works for jobs and not loadavg. But both should show up with no arguments. Only one does for some reason.

**Comment by wltjr** [comment]
I tried with false instead of zero. Same thing, works for jobs but not loadavg. I will see if swapping out the code @bcoca provided, and see if that makes a difference. Really odd that loadavg does not show with either zero or false, for no arguments.

**Comment by webknjaz** [comment]
@wltjr I'm going to guess that you might be testing not the latest version. Let me check this as well. I only checked that it works for `False`. And `0.0` I think. Let's see..

P.S. It seems I've missed an interesting conversation about feature-branch approach here. Just to add my 5 cents :)
If you take into account the overall philosophy of git, it encourages people to preserve history, *pull and merge* changes from each other (which is opposite to the current GitHub flow, where forks send PRs to upstream). Not rewriting history *helps you* to track the flow of author doing decisions. Here's a more verbose explanation of why squashing is considered harmful: https://blog.jaraco.com/dont-use-squash-and-merge/
Another post explaining non-corporate approach: https://dpc.pw/blog/2017/08/youre-using-git-wrong/

OTOH because of certain restrictions of the project and git being [multiparadigm/multiworkflow](https://git-scm.com/book/en/v2/Distributed-Git-Distributed-Workflows) thing it's been chosen to do squashes. I'm not saying that it's wrong or so, but in some cases I'd insist on forbidding squashes.

FTR Ansible has "Create a merge commit" button disabled in repo settings:
screenshot from 2018-02-19 19-46-01
So you may not worry about getting merge commits in trunk ;)

**Comment by wltjr** [comment]
@webknjaz it was just how I was viewing, adding [.patch](https://github.com/ansible/ansible/pull/36355.patch) vs [.diff](https://github.com/ansible/ansible/pull/36355.diff). Viewing diff makes it all one and easier to review. That makes it being 1 or many commits moot. I get your logic. Just was trying to see total changes all together. It was hard to follow some of the pieces with changes.

I am definitely using the very latest from git. You can see [my fork is even with ansible:devel](https://github.com/Obsidian-StudiosInc/ansible) The branch is based off that.

I have tried a few modifications of the code. For some reason loadavg will not show up. jobs works fine, and I can definitely see the change there. Without the changes, no way to get jobs without any arguments. That shows up and works. But for some reason is not working for loadavg. I tried with setting loadavg to int, and that did not make any difference. Trying lots of options and nothing seems to make a difference....

**Comment by webknjaz** [comment]
@wltjr I've downloaded the module from this URL (which corresponds to the last commit) https://raw.githubusercontent.com/webknjaz/ansible/78d5c138f5dbd280264b4152704a131bb5dff6e6/lib/ansible/modules/packaging/os/portage.py
Then, I've put it into `./library/` next to my playbook. I couldn't reproduce your issue on my laptop. Used the following for test values: `4`, `False`, `0`, `0.0`. And everything works as expected.

**Comment by wltjr** [comment]
I am testing an actual install. I updated [my ansible ebuild](https://github.com/Obsidian-StudiosInc/os-xtoo/tree/master/app-admin/ansible) to pull from my repo via snapshot.

```diff
+elif [[ ${PV} == *36355 ]]; then
+       MY_SNAP="649cd61a174f9411a054ef30c07a63b20dae133d"
+       SRC_URI="https://github.com/Obsidian-StudiosInc/ansible//archive/${MY_SNAP}.tar.gz -> ${MY_P}.tar.gz"
+       KEYWORDS="~amd64"
+       MY_S="${PN}-${MY_SNAP}"

```


Maybe something in my env is causing the problem. I noticed in verbose output it is there, but not showing in the remote host.
```yaml
ok: [wlt] => {
    "changed": false,
    "cmd": [
        "/usr/bin/emerge",
        "--getbinpkg",
        "--update",
        "--jobs",
        "--load-average",
        "--ask=n",
        "nano",
        "curl"
    ],
```

**Comment by webknjaz** [comment]
I even removed jinja rendered value from my play (originally put var separately) and still cannot reproduce the issue.

**Comment by webknjaz** [comment]
Do you use a variable there? is it possible that it gets overridden from some unexpected place? (role defaults have the lowest precedence, for example)

**Comment by wltjr** [comment]
@webknjaz I just pushed an [updated ebuild](https://github.com/Obsidian-StudiosInc/os-xtoo/blob/master/app-admin/ansible/ansible-9999.ebuild) if you want to grab it and merge directly. Just grab the raw 9999 and save as the version of the symlink, ansible-2.5.0_beta36355.ebuild. Then you can run the playbook, just change the host to another. Thinking something in my env is causing this issue if it is not happening for you.

I was going to sign off before testing. I am not sure if my testing has made it worse or not...

**Comment by wltjr** [comment]
If I set ```loadavg: True``` I get ```--load-average=1.0```. When set to false or zero, or even -1, it never shows up.  Seems to have something to do with float vs int. But I have changed it to int and it still did not show.

**Comment by wltjr** [comment]
@webknjaz I think your modifications are good. I feel like something else is going wrong either in ansible or the module. Seems something is eating it, as it shows in verbose output, but the command never shows on the actual system. Not sure what would cause such difference. Or why only True makes it show up for me... I am not a python fan really this is not helping change that....

**Comment by webknjaz** [comment]
I'll think about it and maybe do additional checks tomorrow (left my laptop at office). BTW what are the versions of pythons that you run on controller and host? I don't think they influence behavior with number conversion, but still let's check this as well. I was running controller under Python 3.6.4 with `--connection=local`, but it likely picked system's 3.5 for running module.
Also, I can tell you where to put dirty debug hack to check vars in runtime.

**Comment by wltjr** [comment]
@webknjaz no worries, just being overly cautious. Since it was not really ideal the first time around, which you are addressing. I am on python 3.5, dev-lang/python-3.5.4-r1:3.5/3.5m::gentoo. I can see about moving to 3.6, I had some issues due how Gentoo handles such.

I am open to any hacks you can suggest. I tried putting some in unsuccessfully. But it ilkely needs to be further down or something. Since at one point the command is correct, it has the empty --load-average, but it never makes it to the command that is run. Its either lost somewhere further in module code, or some bug in Ansible maybe. Hopefully not that as it would be much deeper. Need to check the final output of the module to see if the argument is there, or not. If it is not then lost in module, if it is then lost in ansible...

Either way thanks again for the contribution and working on this. Seems I did not do a good job when first adding these options. No worries we will get it right this time. At some point...

**Comment by wltjr** [comment]
Ok we are good to go here. This is some odd portage bug not ansible or portage module. If you put ```--load-average``` with no args on the command line, it does not show in the log output. ```--jobs``` shows up regardless, with or without arguments, but not ```--load-average```. That will only show with options. Likely another portage bug, as it should likely be logging/outputting the exact commands.

It explains why ansible verbose was showing. I even exploded the module and it was in the args. Which led me to look into if it would show up on cli when I typed it, and it did not. Thus we are good here.

**Comment by wltjr** [comment]
shipit

**Comment by wltjr** [comment]
I opened a couple related Gentoo bugs if interested

Add --load-average=x.x to man page
https://bugs.gentoo.org/648064

Add --load-average without args to command logged in emerge.log
https://bugs.gentoo.org/648202

**Comment by webknjaz** [comment]
@wltjr ah! I was checking ansible verbose, which shown me actual command it invoked, but you looked into portage's own log :)

**Comment by webknjaz** [comment]
@wltjr @bcoca Thank you for review!
